controllers/gateway/apirule_controller.go

```diff
@@ -285,6 +269,50 @@ func handleDependenciesError(name string, err error) controllers.Status {
 	}
 }
 
+func discoverGateway(client client.Client, ctx context.Context, l logr.Logger, rule *gatewayv2alpha1.APIRule) (*networkingv1beta1.Gateway, bool, error) {
+	if rule.Spec.Gateway == nil {
+		return nil, false, fmt.Errorf("expected Gateway to be set")
+	}
+
+	// The regex pattern is the exact same as the one used in the APIRule validation
+	// Unfortunately usage of a constant is not possible here, as the Kubebuilder CRD validation interface is comment based
+	match, err := regexp.MatchString(`^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?/([a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?)$`, *rule.Spec.Gateway)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if !match {
+		return nil, false, fmt.Errorf("expected Gateway %s to be in the namespace/name format", *rule.Spec.Gateway)
+	}
+
+	gatewayName := strings.Split(*rule.Spec.Gateway, "/")
+	gatewayNN := types.NamespacedName{
+		Namespace: gatewayName[0],
+		Name:      gatewayName[1],
+	}
+	var gateway networkingv1beta1.Gateway
+	if err := client.Get(ctx, gatewayNN, &gateway); err != nil {
+		v2Alpha1Status := status.ReconciliationV2alpha1Status{
+			ApiRuleStatus: &gatewayv2alpha1.APIRuleStatus{
+				State: gatewayv2alpha1.Error,
+			},
+		}
+		s := v2Alpha1Status.GenerateStatusFromFailures([]validation.Failure{
+			{
+				AttributePath: "spec.gateway",
+				Message:       "Could not get specified Gateway",
+			},
+		})
+		if err := s.UpdateStatus(&rule.Status); err != nil {
+			l.Error(err, "Error updating APIRule status")
+			return nil, false, err
+		}
+		return nil, true, err
```

Comment:
can we get rid of `convertStatus` return parameter? having 3 kind of increase complexity, reduce readability and can be case-matched by other two return parameters, e.g. `true` if `return nil, nil` happens and `false` otherwise.